synchronicity/genstub.py

```diff
@@ -0,0 +1,131 @@
+import inspect
+import types
+from unittest import mock
+
+class ReprObj:
+    def __init__(self, repr):
+        self._repr = repr
+
+    def __repr__(self):
+        return self._repr
+
+    def __str__(self):
+        return self._repr
+
+
+
+class StubEmitter:
+    def __init__(self, target_module):
+        self.target_module = target_module
+        self.imports = set()
+        self.parts = []
+        self._indentation = "    "
+
+
+    def formatannotation(self, annotation, base_module=None):
+        assert base_module is None  # don't think this arg is used by signature, but lets check
+        # modified version of the stdlib formatannotations:
+        # if isinstance(annotation, types.GenericAlias):
+        #     return str(annotation)
+        origin = getattr(annotation, "__origin__", None)
+        if origin is None:
+            if isinstance(annotation, type):
+                if annotation == None.__class__:
+                    return "None"
+                if annotation.__module__ in ('builtins', self.target_module):
+                    return annotation.__qualname__
+                return annotation.__module__+'.'+annotation.__qualname__
+            return repr(annotation)
+        # generic:
+        args = getattr(annotation, "__args__", ())
+        return annotation.copy_with(tuple(ReprObj(self.formatannotation(arg)) for arg in args))
+
+    def indent(self, level):
+        return level * self._indentation
+
+    def _get_function(self, func, indentation_level=0):
+        # return source code of function and track imports
+        for annotation in func.__annotations__.values():
+            self._type_imports(annotation)
+
+        return self._get_func_stub_source(func, indentation_level)
+
+    def add_function(self, func, indentation_level=0):
+        # adds function source code to module
+        self.parts.append(self._get_function(func, indentation_level))
+
+    def _import_module(self, module: str):
+        if module not in (self.target_module, "builtins"):
+            self.imports.add(module)
+
+    def _type_imports(self, type_annotation):
+        origin = getattr(type_annotation, "__origin__", None)
+        if origin is None:
+            # "scalar" base type
+            self._import_module(type_annotation.__module__)
+            return
+
+        self._import_module(type_annotation.__module__)  # import the generic itself's module
+        for arg in getattr(type_annotation, "__args__", ()):
+            self._type_imports(arg)
+
+    def get_source(self):
+        import_src = "\n".join(sorted(f"import {mod}" for mod in self.imports))
+        stubs = "\n\n".join(self.parts)
+        return f"{import_src}\n\n{stubs}".lstrip()
+
+    def _get_func_stub_source(self, func, indentation_level):
+        async_prefix = ""
+        if inspect.iscoroutinefunction(func) or inspect.isasyncgenfunction(func):
```

Comment:
can also happen if it's a non-async function returning a coroutine

Reply:
Yeah but those need to be annotated and the annotation will stay in the function source here, so they shouldn't have the `async` prefix in the stub generation as that would add another layer of `asyncness`. We could of course repackage them and unpack the awaitable (this is done for blocking code in the transformation layer already though if that's what you are referring to?)